### PR TITLE
ARROW-2007: [Python] Implement float32 conversions, use NumPy dtype when possible for inner arrays

### PR DESCRIPTION
**Pull request by wesm**

### CODE REVIEW

**Comment by wesm** [comment]
This build is passing save for an apt flake in a java build. @cpcloud or @pitrou could you have a look at this?

File: python/pyarrow/tests/test_array.py [hunk 1]
```diff
@@ -507,6 +507,19 @@ def test_array_from_numpy_datetimeD():
...
+def test_array_from_py_float32():
+    data = [[1.2, 3.4], [9.0, 42.0]]
+    arr1 = pa.array(data[0], type=pa.float32())
+    arr2 = pa.array(data, type=pa.list_(pa.float32()))
+
+    expected1 = np.array(data[0], dtype=np.float32)
+    expected2 = pd.Series([np.array(data[0], dtype=np.float32),
+                           np.array(data[1], dtype=np.float32)])
+
+    assert arr1.equals(pa.array(expected1))
```
Comment (pitrou) [inline python/pyarrow/tests/test_array.py]: Do we want to check that `arr1.type` is exactly `pa.float32()` here?

Comment (wesm) [reply python/pyarrow/tests/test_array.py]: That's included in the call to `equals`, but I'll add an explicit check

**Comment by wesm** [comment]
Shoot, need to pin Appveyor also. @xhochy I'm handling the thrift-cpp pinning here, feel free to merge once appveyor is OK

**Comment by wesm** [comment]
Passing appveyor build on that last commit: https://ci.appveyor.com/project/wesm/arrow/build/1.0.1649. Merging

**Pull request closed without merging**